src/detect.rs

```diff
@@ -18,3 +23,196 @@ pub enum DetectOutcome {
     Pass(BuildPlan),
     Fail,
 }
+
+enum BuildpackTomlArg<BM> {
+    Path(PathBuf),
+    Str(String),
+    BuildpackToml(BuildpackToml<BM>),
+}
+
+enum PlatformArg<P> {
+    Plat(P),
+    Dir(PathBuf),
+}
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
```

Comment:
We could model this without `Option` by creating an enum with two options: `Temporary` and `Static(PathBuf)`. In terms of code this is very similar to using an `Option`, but with additional semantics.

Reply:
I think we need three states. "user set path", "user says to use a temporary directory" and "user didn't say". 

Right now we're depending on injection of a temporary directory. If we want the builder to generate the temporary directories then it would also need to retain a reference to the original tempdir to avoid it being cleared right away. I brought this up in https://github.com/Malax/libcnb.rs/pull/93#issuecomment-925815869

With that API it would execute in a oneliner like this, but it's not clear that the reference to the builder is actually gone and the tempdirs would be cleaned up:

```
let detect_context = DetectContextBuilder::new("heroku-20")
    .buildpack_toml_from_file("../buildpack.toml")
    .with_temp_dirs()
    .build()
    .unwrap()
    .context;
```

I wouldn't want to default to `Temporary` either since in the case where we're using this in `runtime.rs` for example, I want it to error early if no app dir is set by accident.

Reply:
I LOVE this approach. I refactored all the options out to custom enums. It makes the match logic really clean and simplifies the "setter" methods.